yt_dlp/extractor/jiosaavn.py

```diff
@@ -1,93 +1,120 @@
+            'id': 'IcoLuefJ',
             'ext': 'm4a',
             'title': 'Leja Re',
             'album': 'Leja Re',
             'thumbnail': 'https://c.saavncdn.com/258/Leja-Re-Hindi-2018-20181124024539-500x500.jpg',
             'duration': 205,
             'view_count': int,
             'release_year': 2018,
-            'artists': ['Sandesh Shandilya', 'Dhvani Bhanushali', 'Tanishk Bagchi', 'Rashmi Virag', 'Irshad Kamil'],
+            'artists': ['Sandesh Shandilya', 'Dhvani Bhanushali', 'Tanishk Bagchi'],
+            '_old_archive_ids': ['jiosaavnsong OQsEfQFVUXk'],
         },
     }, {
         'url': 'https://www.saavn.com/s/song/hindi/Saathiya/O-Humdum-Suniyo-Re/KAMiazoCblU',
         'only_matching': True,
     }]
 
-    _VALID_BITRATES = ('16', '32', '64', '128', '320')
-
     def _real_extract(self, url):
-        audio_id = self._match_id(url)
-        extract_bitrates = self._configuration_arg('bitrate', ['128', '320'], ie_key='JioSaavn')
-        if invalid_bitrates := [br for br in extract_bitrates if br not in self._VALID_BITRATES]:
-            raise ValueError(
-                f'Invalid bitrate(s): {", ".join(invalid_bitrates)}. '
-                + f'Valid bitrates are: {", ".join(self._VALID_BITRATES)}')
-
-        song_data = self._extract_initial_data(url, audio_id)['song']['song']
-        formats = []
-        for bitrate in extract_bitrates:
-            media_data = self._download_json(
-                'https://www.jiosaavn.com/api.php', audio_id, f'Downloading format info for {bitrate}',
-                fatal=False, data=urlencode_postdata({
-                    '__call': 'song.generateAuthToken',
-                    '_format': 'json',
-                    'bitrate': bitrate,
-                    'url': song_data['encrypted_media_url'],
-                }))
-            if not media_data.get('auth_url'):
-                self.report_warning(f'Unable to extract format info for {bitrate}')
-                continue
-            ext = media_data.get('type')
-            formats.append({
-                'url': media_data['auth_url'],
-                'ext': 'm4a' if ext == 'mp4' else ext,
-                'format_id': bitrate,
-                'abr': int(bitrate),
-                'vcodec': 'none',
-            })
-
-        return {
-            'id': audio_id,
-            'formats': formats,
-            **traverse_obj(song_data, {
-                'title': ('title', 'text'),
-                'album': ('album', 'text'),
-                'thumbnail': ('image', 0, {url_or_none}),
-                'duration': ('duration', {int_or_none}),
-                'view_count': ('play_count', {int_or_none}),
-                'release_year': ('year', {int_or_none}),
-                'artists': ('artists', ..., 'name', {str}, all, {orderedSet}),
-            }),
-        }
+        song_data = self._extract_initial_data(url, self._match_id(url))['song']['song']
+        return self._extract_song(song_data)
```

Comment:
A pattern like this would remove the need for the bool parameter in `self._extract_song`:
```suggestion
        song_data = self._extract_initial_data(url, self._match_id(url))['song']['song']
        return {
            **self._extract_song(song_data),
            'formats': list(self._extract_formats(song_data)),
        }
```

Reply:
Also see discussion on discord regarding not extracting the data twice

Reply:
originally wrote this PR so that the playlist extractors yielded video results instead of url results, but I didn't like having to implement option parsing logic for `extract_flat` inside the extractor, so made a few tweaks. due to `js_to_json` adding 10 seconds every time I think I'll go with the `smuggle_url`/`url_transparent` results impl even if i don't like `smuggle_url`